### ISSUE TITLE
Foam / wake vibration when approaching whirlpools

### ISSUE THREAD

**Issue by Revmatek** [wave-harmonic/crest#1123]
### Is there an existing issue for this?

- [X] I have searched the existing issues.

### Have you checked the documentation to resolve your problem?

- [X] I have read the linked documentation and could not resolve my issue.

### Current Behavior

If a boat or other object approaches a whirlpool (in various configurations) as the foam or wake begins to touch the spiral / flow the rendering of the foam will beginning shake and / or vibrate from its current position as it is sucked into the flow of the whirlpool

### Expected Behavior

As the boat wake approaches the whirlpool and is sucked it, the foam should appear to move smoothly with the flow of the whirlpool as it is sucked in.

### Steps To Reproduce

1. Open the "BoatWakes" dev scene
2. Add an empty gameobject to the scene (ensure that it is in the path of the sphere wakes)
3. Add the "Crest Whirlpool" script to the empty gameobject (increasing the radius may better display the issue)
4. Hit the play button
5. Move the camera to observe the foam being generated as it enters the flow of the whirlpool from various angles
6. Observe jitter / shaking in some cases
Note: The shakiness may not always be immediately apparent but it can be much much worse at times. Given this image below (obviously not shaking because it is a still frame) the vibration in the render occurs at the forward most edge of the foam on the right. Depending on how things are configured the entire foam cap inside the slow will vibrate
<img width="326" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/5d14e995-c497-4f45-86c7-42c6b57b827a">

### Unity Version

2021.3.26f1

### Crest Version

4.17.3

### Render Pipeline

Built-In

### Editor or Standalone

Editor (Edit Mode), Editor (Play Mode), Standalone

### Environment

```markdown
- OS: Windows
```

### Anything else?

I am fairly certain this used to work in an older version (4.15.2 maybe), I recently updated my version and resolved many of the other issues I had mentioned previously. I was unable to mitigate this behavior by tweaking the settings of the whirlpool other than setting the eye radius to 0 which is not really a solution.

**Comment by daleeidd** [comment]
Flow does cause pulsing artefacts. Is that what you are seeing? I will need a video to understand if it is not too much trouble.

**Comment by Revmatek** [comment]
> Flow does cause pulsing artefacts. Is that what you are seeing? I will need a video to understand if it is not too much trouble.

I will get some video.

**Comment by Revmatek** [comment]
> Flow does cause pulsing artefacts. Is that what you are seeing? I will need a video to understand if it is not too much trouble.

Videos are low res and broken up to support upload.

This first one is the scene more or less following the repro steps:

https://github.com/wave-harmonic/crest/assets/114962753/c5df8383-90fb-4f7e-ac70-3c1015e725f7

https://github.com/wave-harmonic/crest/assets/114962753/9341f332-1041-4625-bb1b-e6f464746ea5


This one is with an actual physics objects / boat stuff happening.

https://github.com/wave-harmonic/crest/assets/114962753/7245c05a-9a1a-42c4-9377-01c01d05ac77



This is something new I am pretty sure, it did not used to do this.

**Comment by daleeidd** [comment]
Thanks. I wasn't able to reproduce it myself. Not sure what it could be.

Have you changed any of the simulation settings files (Foam or Dynamic Waves)? There is a simulation frequency setting for both that may help.

What is your frame rate?

**Comment by Revmatek** [comment]
The framerate for these in game is ~60FPS. I believe the framerate in the example scene is unbound so it was probably ~300fps. The only settings I changed in the boatwakes scene was zeroing out the spectrum for the waves to make it easier to see and adding whirlpools on empty game objects where I moved around and changed the radius.

I was able to verify this is some sort of regression, I ran the same test with the BoatWakes scene in crest version 4.15 (same unity version) and everything is nice and smooth as it gets sucked down into the whirlpools. It seems reproducible on multiple computers with 4.17.

Here is some more video comparing the 2 versions. As far as I can tell both of these have the same settings and are unaltered from the branch, with the exception of the spectrum and addition of whirlpools. The only different I noticed is that the whirlpools do not appear until playback has started in 4.15

4.15 (smooth):

https://github.com/wave-harmonic/crest/assets/114962753/a57bbc12-426a-4da2-8cf3-2804e62ec036

4.17.3 (shaky):
<img width="1026" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/9ef59cfd-f656-4f7b-a0f3-d565fc4ad443">

https://github.com/wave-harmonic/crest/assets/114962753/48199652-ca1d-48cf-bfbd-f795c063a57e

**Comment by Revmatek** [comment]
>There is a simulation frequency setting for both that may help.

For both 4.15 and 4.17 BoatWake scenes the simulation frequency is:
<img width="447" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/cdd760c7-f41f-413d-96de-578d987c23a8">

Increasing this value in 4.17 has no effect (probably undesirable performance impacts). Decreasing the value seems to make the problem much much worse. Maybe that can help reproduce it. Decreasing this value on 4.15 does degrade visuals but the overall movement is still smooth.

**Comment by Revmatek** [comment]
I'm looking at this more to see if I can help pinpoint the cause. 

I am seeing a fundamental difference in wake foam between versions as well, while not really a concern it may indicate the cause.

4.15
<img width="482" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/3f17ce3e-0d5b-441a-8ee3-6aa3250517c9">
<img width="241" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/a2ae53b7-ef7e-4718-8857-a272eeab80b9">


4.17
<img width="553" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/0b1dcc90-1cdc-4a93-ad2c-726b0aca7557">
<img width="251" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/fb6e8a31-7c27-4767-98d5-137aa9569db9">

**Comment by Revmatek** [comment]
> Thanks. I was able to reproduce it only intermittently. Some machines were harder than others. I was able to reproduce a more subtle jitter reliably 4.17 and 4.15. Not sure if that means 4.15 will reproduce the less subtle case or not.

That is good news, thanks. I'm not concerned with 4.15 except as a comparison. 4.17 fixed a lot of other issues so I'm probably going to stick with it and continue updating as needed. I think there is subtle jitter in 4.15 in general, but not as extreme (that I noticed.) and it did not appear in any game builds.
 
> Are you able to reproduce the more noticeable jitter case (like the red plane video) reliably?

Yes. Its 100% of the time in the current build with 4.17. The BoatWake scenes is about 90% of the time. I am happy to try out any available fixes / branches in the future and / or get it thoroughly tested.
 
> I agree that it may have something to do with SWI and flow together. Since SWI queries flow I think it has to do with queries. I will see if I can get a reliably reproducible together at some point.
👍

**Comment by daleeidd** [comment]
> I'm not concerned with 4.15 except as a comparison

Yeah, it is more about getting a more reproducible case. I am not having much luck and starting to doubt previous findings.

> For both 4.15 and 4.17 BoatWake scenes the simulation frequency is:

Is that Dynamic Waves? Foam also has a simulation frequency. Have you tried increasing it to 120 too? Everything seems related to simulation frequency so far for me. This will be especially the case at high speeds.

**Comment by Revmatek** [comment]
I looked at this some more and I noticed something odd and something else useful. 

What was odd was that after closing the crest project and opening it later (today), it was much harder to see any jitter. I had to play around with a bit to even see a hint of what I saw before. 

I noticed when I was trying to figure out where there jitter went that by removing the camera from the wake rig, the jitter seemed less, so I used a free moving camera and placed some spheres around a new whirlpool and everything seemed mostly fine. So I checked in game and it does seem related to movement in some way. I am going to investigate if I can adjust scripts to see if there jitter goes away, because it does not occur when boats are kinematic.

**Comment by Revmatek** [comment]
If you add a sphere to the demo scene and hardcode the framerate to not be unbound it seems like it is a lot more reproducible and looks very close to what I am seeing in game. In game I have vsync on and get around 60fps so its not exactly the same.

QualitySettings.vSyncCount = 0;
Application.targetFrameRate = 50
<img width="309" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/5d036571-9038-41f1-a06e-4c4faa0c8b63">




Setting the framerate somewhere between 30-60 seems to display the issue. It does not always show though... which is strange. Sometimes I had to change the framerate from 30 to 50 then let it compile then hit play and it happens, other times I set it to 30 and hit play and it seems smooth, but it is definitely easier to happen. It may also help to follow the sphere with the camera as it moves. Hardcoding this behavior in 4.15 does not seem to cause this problem.

Visual reference:
<img width="732" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/5199d794-056a-4786-aecd-b6e648045fff">

Whirlpool settings:
<img width="450" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/426fe3d9-67d3-4663-9b8f-93885caa272d">


Sphere settings:
<img width="352" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/7650663c-23b4-44de-97d4-826d3468ee9f">
<img width="453" alt="image" src="https://github.com/wave-harmonic/crest/assets/114962753/80ccf323-84fd-448b-b4d6-d61f312fd072">

Hope this helps!

**Comment by Revmatek** [comment]
@daleeidd Were you able to reproduce this?

**Comment by Revmatek** [comment]
Here is a repro scene modifying the BoatWakes scene. Move the camera around and follow the cube. Its a bit goofy setup. 

[repro.zip](https://github.com/wave-harmonic/crest/files/12841412/repro.zip)

**Comment by daleeidd** [comment]
I don't believe I was able to reliably reproduce it before, but I could with your scene.

Just to confirm the issue, is it only related to physics objects? Applying interpolation to the rigid body to the repro seems to be a pretty good workaround.

**Comment by Revmatek** [comment]
> I don't believe I was able to reliably reproduce it before, but I could with your scene.
> 
> Just to confirm the issue, is it only related to physics objects? Applying interpolation to the rigid body to the repro seems to be a pretty good workaround.

Excellent, glad you are able to see the issue now. 

It does seem directly related to being tied to a physics object and flow/foam. PhysX Interpolation is not an option and does not resolve the problems at runtime or in my project. Override foam on "static" objects or transform based movement seems to have less issue.

One bizarre note about this I have been seeing off an on is that solutions can be inconsistent, for example interpolation may momentarily appear to solve the issue the same way that reducing the flow speed or changing some other setting will, but it quickly goes back to vibration foam. Similar things happen when adding new foam to the scene or removing them while attempting to troubleshoot, almost seems like it changes the timing for the better momentarily, couldn't say for sure. 

I would say its some sort of timing / update issue that is revealed by FixedUpdate behavior of some sort. Something in the execution order is moving the foam/flow back to where it was or predicting its location erroneously.

**Comment by Revmatek** [comment]
Any updates on this?

**Comment by daleeidd** [comment]
No. I noticed Follow Horizontal Motion is disabled for the foam input. Does enabling change anything?

**Comment by Revmatek** [comment]
> No. I noticed Follow Horizontal Motion is disabled for the foam input. Does enabling change anything?

It does not appear to have an affect.
